### ISSUE TITLE
Reduce the number of API calls Mautic makes to Salesforce for fetch:leads

### ISSUE THREAD

**Issue by isleshocky77** [mautic/mautic#2836]
What type of report is this:

| Q | A |
| --- | --- |
| Bug report? | N |
| Feature request? | N |
| Enhancement? | Y |
## Description:

It looks like we can reduce the number of API calls being made to Salesforce from Mautic when fetching leads. This is important to anyone trying to run it frequently and has restricted API limits. The additional API queries tend up add up fairly quickly.

Below I would imagine data such as organization creation date and the fields should be cached. I also notice that the number of API calls increased with each new contact added which could prove to be problematic and unnecessary.

```
01 \MauticPlugin\MauticCrmBundle\Api\SalesforceApi::getLeads#L203    SELECT CreatedDate from Organization
02 \MauticPlugin\MauticCrmBundle\Api\SalesforceApi::getLeadFields#82 https://cs12.salesforce.com/services/data/v34.0/sobjects/Lead/describe
03 \MauticPlugin\MauticCrmBundle\Api\SalesforceApi::getLeads#219     SELECT LastName, FirstName, Salutation, Title, Company, Street, City, State, PostalCode, Country, Latitude, Longitude, Phone, MobilePhone, Email, Website, Description, LeadSource, Status, Industry, Rating, AnnualRevenue, NumberOfEmployees, IsUnreadByOwner, EmailBouncedReason, EmailBouncedDate, Deskcom__twitter_username__c, Not_Interested__c, Lead_Role_Responsibility__c, NCAA_Division__c, Request_Demo__c, Subscribe_to_Newsletter__c, id from Lead where LastModifiedDate>=2016-10-24T15:47:00+00:00 and LastModifiedDate<=2016-10-26T15:47:00+00:00
04 \MauticPlugin\MauticCrmBundle\Api\SalesforceApi::getLeadFields#82 https://cs12.salesforce.com/services/data/v34.0/sobjects/Lead/describe
05 \MauticPlugin\MauticCrmBundle\Api\SalesforceApi::getLeadFields#82 https://cs12.salesforce.com/services/data/v34.0/sobjects/Lead/describe
06 \MauticPlugin\MauticCrmBundle\Api\SalesforceApi::getLeadFields#82 https://cs12.salesforce.com/services/data/v34.0/sobjects/Lead/describe
07 \MauticPlugin\MauticCrmBundle\Api\SalesforceApi::getLeadFields#82 https://cs12.salesforce.com/services/data/v34.0/sobjects/Lead/describe

08 \MauticPlugin\MauticCrmBundle\Api\SalesforceApi::getLeads#L203    SELECT CreatedDate from Organization
09 \MauticPlugin\MauticCrmBundle\Api\SalesforceApi::getLeadFields#82 https://cs12.salesforce.com/services/data/v34.0/sobjects/Contact/describe
# Contact
10 \MauticPlugin\MauticCrmBundle\Api\SalesforceApi::getLeads#219     SELECT LastName, FirstName, Salutation, OtherStreet, OtherCity, OtherState, OtherPostalCode, OtherCountry, OtherLatitude, OtherLongitude, MailingStreet, MailingCity, MailingState, MailingPostalCode, MailingCountry, MailingLatitude, MailingLongitude, Phone, Fax, MobilePhone, HomePhone, OtherPhone, AssistantPhone, Email, Title, Department, AssistantName, LeadSource, Birthdate, Description, EmailBouncedReason, EmailBouncedDate, Online_Profile_Team__c, Email_Personal__c, Email_Secondary__c, Skype_Name__c, Middle_Name__c, Suffix__c, Nickname__c, CSCS__c, NSCA__c, MIlitary_NIPR_EMail__c, Primary_Accounting_Contact__c, Military_SIPR_EMail__c, Secondary_Accounting_Contact__c, ATC__c, USAW__c, Administrative_Notification__c, CSSS__c, USATF__c, ACSM__c, NASE__c, Deskcom__twitter_username__c, Personal_URL__c, User__c, Password__c, Role_Responsibility__c, id from Contact where LastModifiedDate>=2016-10-24T15:47:00+00:00 and LastModifiedDate<=2016-10-26T15:47:00+00:00
# Lead
11 \MauticPlugin\MauticCrmBundle\Api\SalesforceApi::getLeads#219     SELECT LastName, FirstName, Salutation, OtherStreet, OtherCity, OtherState, OtherPostalCode, OtherCountry, OtherLatitude, OtherLongitude, MailingStreet, MailingCity, MailingState, MailingPostalCode, MailingCountry, MailingLatitude, MailingLongitude, Phone, Fax, MobilePhone, HomePhone, OtherPhone, AssistantPhone, Email, Title, Department, AssistantName, LeadSource, Birthdate, Description, EmailBouncedReason, EmailBouncedDate, Online_Profile_Team__c, Email_Personal__c, Email_Secondary__c, Skype_Name__c, Middle_Name__c, Suffix__c, Nickname__c, CSCS__c, NSCA__c, MIlitary_NIPThe more details the better...
R_EMail__c, Primary_Accounting_Contact__c, Military_SIPR_EMail__c, Secondary_Accounting_Contact__c, ATC__c, USAW__c, Administrative_Notification__c, CSSS__c, USATF__c, ACSM__c, NASE__c, Deskcom__twitter_username__c, Personal_URL__c, User__c, Password__c, Role_Responsibility__c, id from Contact where LastModifiedDate>=2016-10-24T15:47:00+00:00 and LastModifiedDate<=2016-10-26T15:47:00+00:00
12 \MauticPlugin\MauticCrmBundle\Api\SalesforceApi::getLeadFields#82 https://cs12.salesforce.com/services/data/v34.0/sobjects/Contact/describe
```

**Comment by dwaynetaylor** [comment]
I would add to this, filters on contacts/leads/objects (on both criteria of CRM and Mautic) to pull and push in the API to reduce the calls. This could be overall against the account or at a contact/object specific level via a field being flagged.
Currently deleting contacts which have 65 duplicates (400 x 65 = 26000 duplicate contacts) as a result of no email address and a never ending loop of API calls, this has maxed out the API Limit on Salesforce (15,000 calls per day) as it attempts to sync duplicate contacts.

I am hoping a new field mapping of Dedup on Salesforce ID will prevent this from happening again, but ultimately I have no need for these contacts in digital marketing automation tool as they have no digital footprint.

**Comment by isleshocky77** [comment]
@dwaynetaylor It looks like @mqueme is trying to address part of your issue in #2929

**Comment by isleshocky77** [comment]
@escopecz This is more of an "Feature Enhancement" than a "Feature Request".

**Comment by isleshocky77** [comment]
This issue was addressed by #2954 which was merged and released as part of milestone 2.6.0
